lib/rspec/matchers/built_in/be_kind_of.rb

```diff
@@ -5,10 +5,67 @@ module BuiltIn
       # Provides the implementation for `be_a_kind_of`.
       # Not intended to be instantiated directly.
       class BeAKindOf < BaseMatcher
+        def initialize(*)
+          super
+          @no_checkers_implemented = false
+        end
+
+        # @api private
+        # @return [Boolean]
+        def matches?(actual)
+          @actual = actual
+          result = match(expected, actual)
+          @no_checkers_implemented ? false : result
+        end
+
+        # @api private
+        # @return [Boolean]
+        def does_not_match?(actual)
+          @actual = actual
+          result = match(expected, actual)
+          @no_checkers_implemented ? false : !result
+        end
+
+        # @api private
+        # @return [String]
+        def failure_message
+          if @no_checkers_implemented
+            warning_message
+          else
+            super
+          end
+        end
+
+        # @api private
+        # @return [String]
+        def failure_message_when_negated
+          if @no_checkers_implemented
+            warning_message
+          else
+            super
+          end
+        end
+
       private
 
         def match(expected, actual)
-          actual.kind_of? expected
+          if actual_object_respond_to?(actual, :kind_of?)
+            actual.kind_of?(expected)
+          elsif actual_object_respond_to?(actual, :is_a?)
+            actual.is_a?(expected)
+          else
+            @no_checkers_implemented = true
```

Comment:
Rather than storing the result of the check, I suggest that you raise an `ArgumentError` here with the warning message and shortcut the rest.